### DISCUSSION TITLE
Connecting to rocket server is really slow

### DISCUSSION THREAD

**Discussion by BismuthGlass** [rwf2/Rocket#2766] [Questions] [answered]
Hello,

Some context: I'm using windows, rocket 0.5.

I have noticed significant delays when making requests to my local Rocket server using CURL, with each GET request taking upwards of 250ms. For comparison, GETting `http://www.google.com` with CURL consistently takes me about 100ms. I have created an "empty" rocket server to test this with, and the problem persists. Here is the code, lifted directly from Rocket's docs:

```
#[macro_use] extern crate rocket;

#[get("/")]
fn index() -> &'static str {
    "Hello, world!"
}

#[launch]
fn rocket() -> _ {
    rocket::build()
        .mount("/", routes![index])
}
```

No extra configs, features, or anything of the sort, yet a CURL request to this simple server consistently takes around 250ms, which is completely unexpected.

My question is, what am I missing here? Is there maybe some fairing or something else that is doing extra checks in the background I'm not aware of? I tried disabling the default Shield fairing with `.attach(Shield::new())`, but that did not make any difference. My best guess is that there is something going on with a TLS handshake attempt that messes everything up, but I don't have TLS enabled in the first place, so I'm not sure if this is right or how I would even fix it.

Any ideas?

**Comment by wezm** [comment]
Are you compiling your project with `--release`?

**Reply by BismuthGlass** [reply]
I have compiled both debug and release builds and I see no difference between the two.

**Reply by wezm** [reply]
Ok, how are you measuring the response time?

**Reply by wezm** [reply]
For reference on my machine (Linux) I see this with your code above:

```
cargo run --release &
wrk http://127.0.0.1:8000
Running 10s test @ http://127.0.0.1:8000
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    29.42us    5.16us 217.00us   70.57%
    Req/Sec   154.31k    11.05k  190.65k    69.31%
  3102295 requests in 10.10s, 733.73MB read
Requests/sec: 307158.22
Transfer/sec:     72.65MB
```

**Reply by BismuthGlass** [reply]
I am using `Measure-Command`:

`Measure-Command {curl http://localhost:8000/}`
`Measure-Command {curl http://google.com/}`

Even if this isn't ideal, I am using the same for the google.com request and getting about half the time:

```
> Measure-Command {curl http://localhost:8000/}
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    13  100    13    0     0     55      0 --:--:-- --:--:-- --:--:--    55

TotalMilliseconds : 244.7793

> Measure-Command {curl http://google.com/}
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   219  100   219    0     0   2386      0 --:--:-- --:--:-- --:--:--  2406

TotalMilliseconds : 116.3359
```

It definitely seems related to setting up the connection with the server, since if I use Invoke-WebRequest, which persists the connection, I get sane times after the first request:

```
> Measure-Command {Invoke-WebRequest http://google.com}                                            

TotalMilliseconds : 306.8917

>Measure-Command {Invoke-WebRequest http://localhost:8000}

TotalMilliseconds : 2073.0762

>Measure-Command {Invoke-WebRequest http://localhost:8000}

TotalMilliseconds : 19.6322
```

I made a request to google.com first, just in case .net needed to initialize something. 2 seconds is pretty wild. 19ms is expected here because Invoke-WebRequest is pretty slow in my experience, and Measure-Command also adds some overhead.

**Reply by SergioBenitez** [reply, marked as answer]
Try not using localhost, which will invoke name resolution. If it does IPv6, it can be very slow on Windows. Use 127.0.0.1.

**Reply by BismuthGlass** [reply]
That's it! Using 127.0.0.1 got rid of the delay. Thank you very much.

**Comment by SergioBenitez** [comment]
You'll need to provide more details about what you're doing. What commands exactly are you running? What log output are they producing? Where (shell environments, WSL, console/terminals) are you running them? Which version of the tools are you using? Can you reproduce the slowdown with another kind of local server (say a simple python -m SimpleHttpServer) or is it just specific to Rocket?
